scripts/build.py

```diff
@@ -684,26 +823,137 @@ def get_recently_changed_filenames(dev_dir, out_dir):
+        build_tasks: deque(Thread). A deque that contains all build tasks queued
+            to be processed.
+    """
+    source_dir = dirnames_dict['dev_dir']
+    staging_dir = dirnames_dict['staging_dir']
+    out_dir = dirnames_dict['out_dir']
+    build_tasks = collections.deque()
+    if not os.path.isdir(staging_dir):
+        # If there is no staging dir, perform build process on all files.
+        print 'Creating new %s folder' % staging_dir
+        build_tasks += generate_build_tasks_to_build_all_files_in_directory(
+            source_dir, staging_dir, file_hashes)
+    else:
+        # If staging dir exists, rebuild all HTML and Python files.
+        mandatory_file_extensions = ('.html', '.py',)
+        print (
+            'Staging dir exists, re-building all %s files'
+            % str(mandatory_file_extensions))
+        mandatory_file_lists = get_filepaths_by_extensions(
+            source_dir, mandatory_file_extensions)
+        build_tasks += generate_build_tasks_to_build_files_from_filepaths(
+            source_dir, staging_dir, mandatory_file_lists, file_hashes)
+        dev_dir_hashes = get_file_hashes(source_dir)
+        print 'Getting files that have changed between %s and %s' % (
+            source_dir, out_dir)
+        recently_changed_filenames = get_recently_changed_filenames(
+            dev_dir_hashes, out_dir)
+        if recently_changed_filenames:
+            print 'Re-building recently changed files at %s' % source_dir
+            build_tasks += generate_build_tasks_to_build_files_from_filepaths(
+                source_dir, staging_dir, recently_changed_filenames,
+                file_hashes)
+            # Clean up files in staging directory that have been removed from
+            # source directory.
+            _execute_tasks(generate_delete_tasks_to_remove_deleted_files(
+                dev_dir_hashes, staging_dir))
+        else:
+            print 'No changes detected. Using previously built files.'
+    return build_tasks
+
+
+def _verify_build(input_dirnames, output_dirnames, file_hashes):
+    """Verify a few metrics after build process finishes:
+        1) Number of files between staging directory and final directory
+        matches.
+        2) The hashes in filenames belongs to the hash dict.
+        3) hashes.js, third_party.min.css and third_party.min.js are built and
+        hashes are inserted.
+
+        Args:
+            input_dirnames: list(str). List of directory paths that contain
+                source files
+            output_dirnames: list(str). List of directory paths that contain
+                built files.
+            file_hashes: dict(str, str). Dictionary with filepaths as keys and
+                hashes of file content as values.
+    """
+    for i in xrange(len(input_dirnames)):
+        # Make sure that all files in source directory and staging directory are
+        # accounted for.
+        _compare_file_count(input_dirnames[i], input_dirnames[i])
```

Comment:
Well ... if you compare the same directory to itself I guess you'll get the same file counts. So I'm not sure what the point of this call is?

Reply:
Whoops, I think I fat-fingered a bunch of copy & paste there...